R/ui_functions.R

```diff
@@ -154,3 +167,140 @@ ui_cp_ki_headcount <- function(country, povline, lkup) {
+                           lkup = lkup)
+    dl <- list(dl)
+    names(dl) <- povline
+  }
+
+  dl <- list(pov_charts = dl)
+
+  # Fetch pre-calculated data (filter selected country)
+  dl2 <- lapply(lkup$cp$charts, function(x) {
+    x[country_code == country]
+  })
+
+  out <- append(dl, dl2)
+
+  return(out)
+
+}
+
+#' Provides numbers that will populate the country profiles poverty charts
+#'
+#' @inheritParams ui_cp_charts
+#' @return list
+#' @keywords internal
+#'
+ui_cp_poverty_charts <- function(country, povline, pop_units,
+                                 year_range, lkup) {
+
+  # Fetch data for poverty trend chart
+  res_pov_trend <-
+    pip(country = country, povline = povline, lkup = lkup)
+  res_pov_trend$pop_in_poverty <-
+    res_pov_trend$reporting_pop * res_pov_trend$headcount / pop_units
+  res_pov_trend <-
+    res_pov_trend[, c('country_code', 'reporting_year',
+                      'headcount', 'pop_in_poverty')]
+
+  # Fetch data for poverty bar chart
+  region <-
+    lkup$svy_lkup[country_code == country]$pcn_region_code %>%
+    unique()
+  countries <-
+    lkup$svy_lkup[pcn_region_code == region]$country_code %>%
+    unique()
+
+  res_pov_mrv <- pip(country = countries, lkup = lkup)
+  res_pov_mrv <-
+    res_pov_mrv[, .SD[which.max(reporting_year)],
+                by = country_code]
+  res_pov_mrv <-
+    res_pov_mrv[reporting_year %in% year_range]
+  res_pov_mrv <-
+    res_pov_mrv[, c('country_code', 'reporting_year',
+                    'poverty_line', 'headcount')]
+  res_pov_mrv <-
+    cp_pov_mrv_select_countries(res_pov_mrv, country)
+
+  # dl1 <- list(res_pov_trend)
+  # names(dl1) <- as.character(povline)
+  # dl2 <- list(res_pov_mrv)
+  # names(dl2) <- as.character(povline)
+
+  out <- list(
+    pov_trend = res_pov_trend,
+    pov_mrv = res_pov_mrv
+  )
+
+  # out <- list(pov_trend = dl1, pov_mrv = dl2)
+
+  return(out)
+
+}
+
+#' Select countries to display for CP Poverty MRV Chart
+#' @param dt data.table: Result from `pip()` query
+#' @param country character: Country code
+#' @noRd
+cp_pov_mrv_select_countries <- function(dt, country){
+  # If the number of countries in the region of the selected country is > 12,
+  # then the countries to be displayed will be selected as follow:
+  if (nrow(dt) > 12) {
+    h <- dt[country_code == country]$headcount
+    v <- sort(dt$headcount)
+    # IF selected country does not pertain to top 5 or bottom 5, display:
+    if (!(h %in% utils::tail(v, 5) | h %in% utils::head(v, 5))) {
+      v <- v[!v %in% h]
+      v2 <- v[v > h]
+      v3 <- v[v < h]
+      vals <- c(h, utils::head(v, 3), utils::tail(v, 3),
+                utils::head(v2, 2), utils::tail(v3, 2))
+      # ELSE IF selected country pertains to top 5, display:
+    } else if (h %in% utils::tail(v, 5)) {
+      vals <- c(utils::head(v, 5), utils::tail(v, 6))
+      # ELSE display
+    } else {
+      vals <- c(utils::head(v, 6), utils::tail(v, 5))
+    }
```

Comment:
I think I would probably extract this whole section into its own function. This is the heart of the selection algorithm, and it would be much easier to unit test edge cases.

Reply:
Also good idea.